src/waiter.c

```diff
@@ -0,0 +1,311 @@
+                goto all_ok;
+            }
+            sm_exec_end(waiter->sm);
+        }
+    }
+all_ok:
+    return result;
+}
+
+static void cancel_notify(void* context)
+{
+    WAITER_HANDLE waiter = context;
+    srw_lock_acquire_exclusive(waiter->lock);
+    {
+        waiter->notify_complete_callback(waiter->notify_complete_callback_context, WAITER_RESULT_CANCELLED);
+        waiter->callbacks_called = true;
+        THANDLE_ASSIGN(RC_PTR)(&waiter->data, NULL);
+        sm_fault(waiter->sm);
+        sm_exec_end(waiter->sm);
+    }
+    srw_lock_release_exclusive(waiter->lock);
+}
+
+THANDLE(ASYNC_OP) waiter_notify(WAITER_HANDLE waiter, THANDLE(RC_PTR) data, NOTIFY_COMPLETE_CALLBACK notify_complete_callback, void* context)
+{
+    THANDLE(ASYNC_OP) result = NULL;
+    if (waiter == NULL || data == NULL || notify_complete_callback == NULL || context == NULL)
+    {
+        LogError("Invalid arguments: WAITER_HANDLE waiter = %p, THANDLE(RC_PTR) data = %p, NOTIFY_COMPLETE_CALLBACK notify_complete_callback = %p, void* context = %p",
+                   waiter, data, notify_complete_callback, context);
+    }
+    else
+    {
+        SM_RESULT sm_exec_begin_result = sm_exec_begin(waiter->sm);
+        if (sm_exec_begin_result != SM_EXEC_GRANTED)
+        {
+            LogError("Failure in sm_exec_begin(waiter->sm=%p)", waiter->sm);
+        }
+        else
+        {
+            bool success;
+            srw_lock_acquire_exclusive(waiter->lock);
+            {
+                if (waiter->notified)
+                {
+                    LogError("waiter_notify called twice on WAITER waiter=%p."
+                        "Previous arguments: THANDLE(RC_PTR) data=%" PRI_RC_PTR ", NOTIFY_COMPLETE_CALLBACK notify_complete_callback=%p, void* context=%p."
+                        "Current arguments: THANDLE(RC_PTR) data=%" PRI_RC_PTR ", NOTIFY_COMPLETE_CALLBACK notify_complete_callback=%p, void* context=%p.",
+                        waiter,
+                        waiter->data, waiter->notify_complete_callback, waiter->notify_complete_callback_context,
+                        data, notify_complete_callback, context
+                    );
+                    success = false;
+                }
+                else
+                {
+                    THANDLE(ASYNC_OP) async_op = async_op_create(cancel_notify, sizeof(WAITER_HANDLE), alignof(WAITER_HANDLE), dispose);
+                    if (async_op == NULL)
+                    {
+                        LogError("Failure in async_op_create(cancel_notify=%p, sizeof(WAITER_HANDLE)=%" PRIu32 ", alignof(WAITER_HANDLE)=%" PRIu32 ", dispose=%p)", cancel_notify, (uint32_t)sizeof(WAITER_HANDLE), (uint32_t)alignof(WAITER_HANDLE), dispose);
+                        success = false;
+                    }
+                    else
+                    {
+                        ((ASYNC_OP*)async_op)->context = waiter;
+
+                        THANDLE_ASSIGN(RC_PTR)(&waiter->data, data);
```

Comment:
>THANDLE_ASSIGN [](http://example.com/codeflow?start=24&length=14)

from all the THANDLE APIs, which one should be used to initialize a variable?

Hint: answer is in the question.

Reply:
Done